File: VisualStudio/CSharp/GUI.cs
﻿using System;
using System.Drawing;
using System.Windows.Forms;

namespace CSharp
{
    public partial class GUI : Form
    {
        private const string APP_NAME = "Tic Tac Toe";
        private int clickCount = 0;
        private bool gameEnded = false;
        private Button[] buttonGrid;

        public GUI()
        {
            InitializeComponent();
            this.exitToolStripMenuItem.Click += (object sender, EventArgs _) => Application.Exit();
            this.restartToolStripMenuItem.Click += RestartToolStripMenu;
            this.button1.Click += GameLogic;
            this.button2.Click += GameLogic;
            this.button3.Click += GameLogic;
            this.button4.Click += GameLogic;
            this.button5.Click += GameLogic;
            this.button6.Click += GameLogic;
            this.button7.Click += GameLogic;
            this.button8.Click += GameLogic;
            this.button9.Click += GameLogic;
            this.Load += GUI_Load;
        }

        private void GUI_Load(object sender, EventArgs e)
        {
            buttonGrid = new Button[9] { button1, button2, button3, button4, button5, button6, button7, button8, button9 };
        }
        private void ResetGame()
        {
            foreach (Button button in buttonGrid)
            {
                button.ResetText();
                button.ForeColor = Button.DefaultForeColor;
            }
            gameEnded = false;
            clickCount = 0;
        }
        private void FinishGame((Button, Button, Button) pattern)
        {
            pattern.Item1.ForeColor = Color.Green;
            pattern.Item2.ForeColor = Color.Green;
            pattern.Item3.ForeColor = Color.Green;
            gameEnded = true;
            if (MessageBox.Show($"Player {(pattern.Item2.Text == "X" ? "one" : "two")} has won!", APP_NAME, MessageBoxButtons.RetryCancel, MessageBoxIcon.Information) == DialogResult.Retry) ResetGame();
        }
        private void RestartToolStripMenu(object sender, EventArgs e)
        {
            if (MessageBox.Show("Are you sure?", APP_NAME, MessageBoxButtons.YesNo, MessageBoxIcon.Question) == DialogResult.Yes) ResetGame();
        }
        private void GameLogic(object sender, EventArgs e)
        {
            if (gameEnded) return;
            Button currentButton = (Button)sender;
            if (currentButton.Text == string.Empty)
            {
                currentButton.Text = clickCount++ % 2 == 0 ? "X" : "O";

                for (int i = 1; i < buttonGrid.Length; i += 3) if ((buttonGrid[i - 1].Text, buttonGrid[i].Text, buttonGrid[i + 1].Text) == ("X", "X", "X") || (buttonGrid[i - 1].Text, buttonGrid[i].Text, buttonGrid[i + 1].Text) == ("O", "O", "O")) FinishGame((buttonGrid[i - 1], buttonGrid[i], buttonGrid[i + 1]));
                for (int i = 3; i < 6; i++) if ((buttonGrid[i - 3].Text, buttonGrid[i].Text, buttonGrid[i + 3].Text) == ("X", "X", "X") || (buttonGrid[i - 3].Text, buttonGrid[i].Text, buttonGrid[i + 3].Text) == ("O", "O", "O")) FinishGame((buttonGrid[i - 3], buttonGrid[i], buttonGrid[i + 3]));
                if ((buttonGrid[0].Text, buttonGrid[4].Text, buttonGrid[8].Text) == ("X", "X", "X") || (buttonGrid[0].Text, buttonGrid[4].Text, buttonGrid[8].Text) == ("O", "O", "O")) FinishGame((buttonGrid[0], buttonGrid[4], buttonGrid[8]));
                else if ((buttonGrid[2].Text, buttonGrid[4].Text, buttonGrid[6].Text) == ("X", "X", "X") || (buttonGrid[2].Text, buttonGrid[4].Text, buttonGrid[6].Text) == ("O", "O", "O")) FinishGame((buttonGrid[2], buttonGrid[4], buttonGrid[6]));
                else if (clickCount == 9 && MessageBox.Show("Draw", APP_NAME, MessageBoxButtons.RetryCancel, MessageBoxIcon.Information) == DialogResult.Retry) ResetGame();
            }
            else MessageBox.Show("You can't modify already modified panels", APP_NAME, MessageBoxButtons.OK, MessageBoxIcon.Exclamation);
        }
    }
}

File: VisualStudio/VisualBasic/GUI.vb
Public Class GUI
    Private Const APP_NAME As String = "Tic Tac Toe"
    Private clickCount As Integer = 0
    Private gameEnded As Boolean = False
    Private buttonGrid(9) As Button

    Private Sub GUI_Load(ByVal sender As Object, ByVal e As EventArgs) Handles Me.Load
        buttonGrid = New Button() {button1, button2, button3, button4, button5, button6, button7, button8, button9}
    End Sub

    Private Sub ResetGame()
        For Each button As Button In buttonGrid
            button.ResetText()
            button.ForeColor = Button.DefaultForeColor
        Next
        gameEnded = False
        clickCount = 0
    End Sub
    Private Sub FinishGame(ByVal pattern As (Button, Button, Button))
        pattern.Item1.ForeColor = Color.Green
        pattern.Item2.ForeColor = Color.Green
        pattern.Item3.ForeColor = Color.Green
        gameEnded = True
        If MsgBox($"Player {IIf(pattern.Item2.Text = "X", "one", "two")} has won!", MsgBoxStyle.RetryCancel + MsgBoxStyle.Information, APP_NAME) = MsgBoxResult.Retry Then ResetGame()
    End Sub
    Private Sub ExitToolStripMenu(ByVal sender As Object, ByVal e As EventArgs) Handles exitToolStripMenuItem.Click
        Application.Exit()
    End Sub
    Private Sub RestartToolStripMenu(ByVal sender As Object, ByVal e As EventArgs) Handles restartToolStripMenuItem.Click
        If MsgBox("Are you sure?", MsgBoxStyle.Question + MsgBoxStyle.YesNo, APP_NAME) = MsgBoxResult.Yes Then ResetGame()
    End Sub
    Private Sub GameLogic(ByVal sender As Object, ByVal e As EventArgs) Handles button1.Click, button2.Click, button3.Click, button4.Click, button5.Click, button6.Click, button7.Click, button8.Click, button9.Click
        If gameEnded Then Return
        Dim currentButton As Button = CType(sender, Button)
        If currentButton.Text = String.Empty Then
            currentButton.Text = IIf(clickCount Mod 2 = 0, "X", "O")
            clickCount += 1

            For i As Integer = 1 To buttonGrid.Count - 1 Step +3
                If (buttonGrid(i - 1).Text = "X" AndAlso buttonGrid(i).Text = "X" AndAlso buttonGrid(i + 1).Text = "X") OrElse (buttonGrid(i - 1).Text = "O" AndAlso buttonGrid(i).Text = "O" AndAlso buttonGrid(i + 1).Text = "O") Then FinishGame((buttonGrid(i - 1), buttonGrid(i), buttonGrid(i + 1)))
            Next
            For i As Integer = 3 To 5
                If (buttonGrid(i - 3).Text = "X" AndAlso buttonGrid(i).Text = "X" AndAlso buttonGrid(i + 3).Text = "X") OrElse (buttonGrid(i - 3).Text = "O" AndAlso buttonGrid(i).Text = "O" AndAlso buttonGrid(i + 3).Text = "O") Then FinishGame((buttonGrid(i - 3), buttonGrid(i), buttonGrid(i + 3)))
            Next
            If (buttonGrid(0).Text = "X" AndAlso buttonGrid(4).Text = "X" AndAlso buttonGrid(8).Text = "X") OrElse (buttonGrid(0).Text = "O" AndAlso buttonGrid(4).Text = "O" AndAlso buttonGrid(8).Text = "O") Then
                FinishGame((buttonGrid(0), buttonGrid(4), buttonGrid(8)))
            ElseIf (buttonGrid(2).Text = "X" AndAlso buttonGrid(4).Text = "X" AndAlso buttonGrid(6).Text = "X") OrElse (buttonGrid(2).Text = "O" AndAlso buttonGrid(4).Text = "O" AndAlso buttonGrid(6).Text = "O") Then
                FinishGame((buttonGrid(2), buttonGrid(4), buttonGrid(6)))
            ElseIf clickCount = 9 AndAlso MsgBox("Draw", MsgBoxStyle.Information + MsgBoxStyle.RetryCancel, APP_NAME) = MsgBoxResult.Retry Then
                ResetGame()
            End If
        Else
            MsgBox("You can't modify already modified panels", MsgBoxStyle.Exclamation + MsgBoxStyle.OkOnly, APP_NAME)
        End If
    End Sub

End Class
